### PR TITLE
Add handler cache for MakeDeepCopy

### PR DESCRIPTION
**Pull request by Mhburg**
The following image is a screenshot from profiling the patch time in RogueTech, a heavily modded version of BattleTech.
As shown in the pic, `MakeDeepCopy` accounts for more than 90% of cpu time. The majority of the time is spent on getting the handler of `Add` method for types.
This PR is a simple optimization on this process by creating a dictionary for fast repetitive accesses.
image

This is the profiling data after the change.
image

Following is a comparison on memory usage,
Before,
image

After,
image

The difference mostly can attribute to transient memory allocation, and as the memory usage shown in the After change, the allocation for persistent memory is quite small. However, it would not hurt if Harmony can open an API for modloaders to clear the cache.

### CODE REVIEW

**Comment by Mhburg** [comment]
Forgot to write, Harmony 1.2 is used in the profiling.
Enviroment is Win10 mono x64

File: Harmony/Tools/AccessTools.cs
Comment (lbmaian) [inline Harmony/Tools/AccessTools.cs:1707]: Does this already handle dictionaries then via `Add(KeyValuePair<...>)`?

File: Harmony/Tools/AccessTools.cs
Comment (lbmaian) [inline Harmony/Tools/AccessTools.cs:1697]: Sneak in another minor optimization here: use `Count` rather than `Count()`

File: Harmony/Tools/AccessTools.cs
Comment (lbmaian) [inline Harmony/Tools/AccessTools.cs:21]: This isn't thread-safe. Ideally we'd use a ConcurrentDictionary, but that isn't supported on net35.

In other areas, we've been using locks as a workaround, but if performance is a concern, then there would have to be two code paths:

net35: uses Dictionary and either:
a) locks on that Dictionary upon usage
b) use ThreadStatic (not ideal for a cache) with manual per-thread initialization

Other: use ConcurrentDictionary

That said, I think always locking is fine, since the dict access performance is minor compared to the addInvoker usage.

**Review by lbmaian — CHANGES_REQUESTED** [review]
Harmony CI is having spurious failures again... I'll keep retrying it manually.

File: Harmony/Tools/AccessTools.cs
Comment (Mhburg) [reply Harmony/Tools/AccessTools.cs:21]: Good to hear that Harmony is moving to multi-thread

Comment (Mhburg) [reply Harmony/Tools/AccessTools.cs:21]: I will switch the code to ConcurrentDictionary

Comment (lbmaian) [reply Harmony/Tools/AccessTools.cs:21]: You can't use ConcurrentDictionary on .NET Framework 3.5 though, which Harmony must still support. This is why I suggest just locking instead.

File: Harmony/Tools/AccessTools.cs
Comment (lbmaian) [reply Harmony/Tools/AccessTools.cs:1697]: Or rather, `Length` since `GetParameters()` returns an array.

**Comment by Mhburg** [comment]
Is preprocessor allowed in Harmony? I can use them for difference version of .NET

On Sep 25, 2020, at 7:56 AM, lbmaian <notifications@github.com> wrote:

﻿

@lbmaian commented on this pull request.

________________________________

In Harmony/Tools/AccessTools.cs<https://github.com/pardeike/Harmony/pull/348#discussion_r494937590>:

> @@ -1687,11 +1692,18 @@ public static object MakeDeepCopy(object source, Type resultType, Func<string, T

                        if (type.IsGenericType && resultType.IsGenericType)
                        {
-                               var addOperation = FirstMethod(resultType, m => m.Name == "Add" && m.GetParameters().Count() == 1);
-                               if (addOperation is object)
+                               if (!handlerCache.TryGetValue(resultType, out FastInvokeHandler addInvoker))
+                               {
+                                       var addOperation = FirstMethod(resultType, m => m.Name == "Add" && m.GetParameters().Count() == 1);


Or rather, Length since GetParameters() returns an array.

**Comment by lbmaian** [comment]
> Is preprocessor allowed in Harmony? I can use them for difference version of .NET

Yes, `#if NET35` blocks already exist in AccessTools.cs. That said, @pardeike wasn't in favor of #if NET35` lock blocks due to code clarity concerns, and decided to just always use locks.

**Comment by Mhburg** [comment]
> > Is preprocessor allowed in Harmony? I can use them for difference version of .NET
> 
> Yes, `#if NET35` blocks already exist in AccessTools.cs. That said, @pardeike wasn't in favor of `#if NET35` lock blocks due to code clarity concerns, and decided to just always use locks.

ok

File: Harmony/Tools/AccessTools.cs
Comment (Mhburg) [reply Harmony/Tools/AccessTools.cs:1697]: It is legacy code, I will make the change.

File: Harmony/Tools/AccessTools.cs
Comment (Mhburg) [reply Harmony/Tools/AccessTools.cs:1707]: that's a legacy comment from Pardeike

File: Harmony/Tools/AccessTools.cs
Comment (lbmaian) [inline Harmony/Tools/AccessTools.cs:1735]: Shouldn't this be `ExitUpgradeableReadLock`?

**Review by lbmaian — CHANGES_REQUESTED** [review]

**Review by lbmaian — APPROVED** [review]
Looks good, will wait for CI to succeed before merging.

**Pull request merged**
